# Puzzle

### This Week’s Fiddler

I’m preparing a mixture of “random-ade” using a large, empty pitcher and two 12-ounce glasses.

First, I fill one glass with some amount of lemon juice chosen randomly and uniformly between 0 and 12 ounces. I fill the other glass with some amount of water, also chosen randomly and uniformly between 0 and 12 ounces. Next, I pour an equal amount from each glass into the pitcher until one of the glasses is empty.

At this point, I refill that empty glass with yet another random amount of the same liquid it previously contained. Once again, I pour an equal amount from each glass into the pitcher until one of the glasses is empty.

On average, how much random-ade can I expect to prepare? (Note that all three random amounts in this problem are chosen independently of each other.)

### This Week’s Extra Credit

Once again I’m preparing random-ade, but this time I have three 12-ounce glasses.

I fill the first glass with a random amount of lemon juice, the second glass with a random amount of lime juice, and the third glass with a random amount of water. As before, each amount is chosen uniformly between 0 and 12 ounces, and all amounts are independent. Next, I pour an equal amount from each glass into the pitcher until one of the glasses is empty.

At this point, I refill that empty glass with yet another random amount of the same liquid it previously contained. Once again, I pour an equal amount from each glass into the pitcher until one of the glasses is empty.

Then I refill that now-empty glass with yet another random amount of the same liquid it previously contained. Again, I pour an equal amount from each glass into the pitcher until one of the glasses is empty.

On average, how much random-ade can I expect to prepare?

# Solution

Just taking the code approach.

In [35]:
def random_ade_2cups(a,b,c):
    glass1, glass2 = a,b
    pour1 = min(glass1, glass2)
    glass1 -= pour1
    glass2 -= pour1
    if glass1 == 0:
        glass1 += c
    else:
        glass2 += c
    pour2 = min(glass1, glass2)
    return 2*(pour1 + pour2) # 2* because we pour that much amount out of every cup, and we have 2 cups

In [36]:
def random_ade_3cups(a,b,c,d,e):
    glass1, glass2, glass3 = a,b,c
    pour1 = min(glass1, glass2, glass3)
    glass1 -= pour1
    glass2 -= pour1
    glass3 -= pour1
    if glass1 == 0:
        glass1 += d
    elif glass2 == 0:
        glass2 += d
    else:
        glass3 += d
    pour2 = min(glass1, glass2, glass3)
    glass1 -= pour2
    glass2 -= pour2
    glass3 -= pour2
    if glass1 == 0:
        glass1 += e
    elif glass2 == 0:
        glass2 += e
    else:
        glass3 += e
    pour3 = min(glass1, glass2, glass3)
    return 3*(pour1 + pour2 + pour3) # 3* because we pour that much amount out of every cup, and we have 3 cups

Now we find the expected value of these functions by integrating over the possible range and dividing by the range size. My hope is that they are a good match for numerical integration because the functions are continuous.

In [37]:
from scipy.integrate import nquad
def my_expected_value(fn, num_dims, min_range=0, max_range=12, eps=1e-6):
    volume = (max_range - min_range) ** num_dims
    def integrand(*args):
        return fn(*args) / volume
    limits = [(min_range, max_range) for _ in range(num_dims)]
    result = nquad(integrand, limits, opts={'epsabs': eps, 'epsrel': eps}, full_output=True)
    print(f"{fn.__name__}: eps={eps} Integral result: {result}")
    return result

In [38]:
for eps in [1e-3, 1e-4, 1e-5, 1e-6]:
    my_expected_value(random_ade_2cups, 3, eps=eps)

random_ade_2cups: eps=0.001 Integral result: (13.999869936371235, 0.0021098698149354493, {'neval': 54285})
random_ade_2cups: eps=0.0001 Integral result: (13.999985724813566, 0.0001261744573178733, {'neval': 346773})
random_ade_2cups: eps=1e-05 Integral result: (13.999993336345316, 1.3034023918594784e-05, {'neval': 1491609})
random_ade_2cups: eps=1e-06 Integral result: (13.999998661019614, 1.317515838919375e-06, {'neval': 2846235})


In [39]:
for eps in [1e-1, 1e-2, 1e-3]:
    my_expected_value(random_ade_3cups, 5, eps=eps)

random_ade_3cups: eps=0.1 Integral result: (22.19922679061998, 0.014934710787150573, {'neval': 19214475})
random_ade_3cups: eps=0.01 Integral result: (22.19922679061998, 0.014934710787150573, {'neval': 19214475})
random_ade_3cups: eps=0.001 Integral result: (22.199501659977365, 0.012683040890676456, {'neval': 51119817})


I think it is pretty safe to speculate at this point that:

* Fiddler answer = 14 ounces.

* Extra Credit answer =  22.2 ounces.